### ISSUE TITLE
performance issue with random.shuffle call in assumptions

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#16265]
In the core assumptions code, there is a call to `random.shuffle`. This was added in ba6e49fa0b7. https://github.com/sympy/sympy/blob/ce64896148f965a117619143aac2c320f7bd3a68/sympy/core/assumptions.py#L298

In trying to profile https://github.com/sympy/sympy/issues/16249, I found that it causes serious slowdowns. In master, with `x = symbols('x:50')` takes 2.3 seconds, but commenting out the shuffle line, it takes 1.75 seconds. The rest of the time is taken from other things, but this is appears to be a serious slowdown, since it's a core part of SymPy that is called many times. 

CC @smichr

**Comment by ShubhamKJha** [comment]
This can be removed. The performance gain is important and the commit [ba6e49f](https://github.com/sympy/sympy/commit/ba6e49fa0b72a9ee93def4aca821cb3b77aaaede) suggests some fail safe to another issue rather than adding anything to the `_ask` function.

**Comment by oscarbenjamin** [comment]
I don't think that random.shuffle particularly slows things down in normal usage. At least that should be demonstrated.

I have found that shuffling the prerequisites can help for detecting inconsistent rules in the assumptions system. The problem manifests itself as non-deterministic behaviour which is awkward but at least will be picked up by test code.

**Comment by ShubhamKJha** [comment]
Hi @oscarbenjamin, can you explain the second part. I couldn't get it from the commit also.

**Comment by oscarbenjamin** [comment]
When you query `is_x` if `_eval_is_x` returns None then prerequisites can be checked. Suppose that y->x and z-->!x then we can check `_eval_is_y` or `_eval_is_z` but which should we check first? This call to shuffle ensures that we don't always try the same thing first so if there is a bug and `_eval_is_y` and `_eval_is_z` are inconsistent with each other then the difference manifests in `is_x` giving non-deterministically different results. This will then show up in testing if you depend on `is_x` giving a particular result.

**Comment by asmeurer** [comment]
I just confirmed the same 1.3x slowdown on another computer for `symbols('x:50')`. Putting some print statements there, the function seems to be called a lot. For instance, calling `sin(x)` caused the shuffle to be called 2642 times (fortunately only the first time because of the cache). The time to create `sin(x)` with the shuffle is 5.73 ms vs. 1.21 without it. That's 4.7x slower.

**Comment by asmeurer** [comment]
I agree with the benefits of shuffling. Can we do the shuffle once, when the prereq dict is generated, rather than in each call? It would be the same order globally so could there still be inconsistencies that aren't caught in that case?

**Comment by ShubhamKJha** [comment]
I don't think there's any benefit other than in testing. Is there any way to have this during testing only?

**Comment by asmeurer** [comment]
I think it's possible. I guess by design this would lead to random test failures, which are only reproducible when reusing the same random seed. With this, they would also only be reproducible in the tests themselves, or when manually enabling the shuffling. So it could be very confusing to someone trying to debug things, unless they were aware this was going on. 

By the way, this is related https://github.com/sympy/sympy/pull/16967

**Comment by oscarbenjamin** [comment]
What do you mean that `sin(x)` is called 2642 times?

I tried this diff:
```diff
diff --git a/sympy/core/assumptions.py b/sympy/core/assumptions.py
index 8e82a24f3c..abf018f45a 100644
--- a/sympy/core/assumptions.py
+++ b/sympy/core/assumptions.py
@@ -313,6 +313,7 @@ def _ask(fact, obj):
 
     # Try assumption's prerequisites
     prereq = list(_assume_rules.prereq[fact])
+    print(obj, fact)
     shuffle(prereq)
     for pk in prereq:
         if pk in assumptions:
```
And this script:
```python
from sympy import *
x = Symbol('x')
sin(x)
```
Running that I get:
```console
$ python x.py 
-1 even
x zero
x extended_negative
x finite
x extended_positive
x positive
x negative
```
I don't see 2642 calls...

Certainly there are some unnecessary calls to `_ask` there though. We can optimise Symbol (which I think has precomputed assumptions) so that `_ask` should never need to call itself recursively like this.

**Comment by asmeurer** [comment]
Oh my bad. I did `sin(x)` in the interpreter, but it looks like most of the calls come from the printing. A lot of it seems to be coming from an initial import of the units module. The performance difference for `a = sin(x)` is negligible. 

I would definitely like to see the actual benchmark suite checked against this. I tested some other functions like `solve` and in some cases it is called a lot. But it is cached, so the true performance impact is hard to judge. It also seems whenever a Dummy symbol is involved it prevents the caching from working.

**Comment by oscarbenjamin** [comment]
Checking assumptions on symbols is a different case from other expressions and could potentially be optimised.

For a pure symbol, dummy or otherwise, there are some assumptions supplied at creation. Other assumptions are inferred from that through `deduce_all_facts`. We should compute those every time a symbol is created (cacheing the result for future symbols) and optimise Symbol so that e.g. `is_positive` is as fast as possible at returning the precomputed fact. Maybe this has already been done - I haven't checked...

Another possibility I've considered is that `_eval_is_` methods could have a 4th return value so it's True, False, None or Unknowable. The significant of Unknowable is that no prerequisites or other inference rules should be checked. This could potentially have speed advantages if used only internally to the assumptions system so that `is_` attributes still return None when `_eval_is` returns Unknowable. Maybe that's just overcomplicating things though.

**Comment by P403n1x87** [comment]
@ShubhamKJha Is there still interest in getting this PR merged?
